src/main/java/fi/aalto/cs/apluscourses/utils/DomUtil.java

```diff
@@ -107,6 +115,39 @@ private DomUtil() {
     }
   }
 
+  /**
+   * Iterates through the immediate children of the given parent and returns the first node with the
+   * given name, attribute, and attribute value, or null if such a node isn't found.
+   * @param parent         The children of this node are examined.
+   * @param tagName        The name of the tag of the child node.
+   * @param attributeName  The name of an attribute that the child node must have.
+   * @param attributeValue The value corresponding to the attribute name that the child must have.
+   */
+  @Nullable
+  public static Node findChildNodeWithAttribute(@NotNull Node parent,
+                                                @NotNull String tagName,
+                                                @NotNull String attributeName,
+                                                @NotNull String attributeValue) {
+    NodeList children = parent.getChildNodes();
+    for (int i = 0; i < children.getLength(); ++i) {
+      Node child = children.item(i);
+      if (!child.getNodeName().equals(tagName)) {
+        continue;
+      }
+
+      NamedNodeMap attributes = child.getAttributes();
+      if (attributes == null) {
+        continue;
+      }
+
+      Node attribute = attributes.getNamedItem(attributeName);
+      if (attribute != null && attribute.getNodeValue().equals(attributeValue)) {
+        return child;
+      }
+    }
+    return null;
+  }
```

Comment:
Could this method be replaced with a simple XPath query?

Reply:
Good idea, done :+1: Although now the method is open to XPath injections, but in this case all input comes from us (i.e. a trusted source), any opinions?

Reply:
After some more refactoring, it now uses a XPath query without any issues